### PR TITLE
33465 Varnish cache adding top menu category ids to all page cache tags

### LINKED ISSUE #33465: Varnish cache adding top menu category ids to all page cache tags
### Preconditions (*)

1. Magento 2.4
2. Enable varnish cache
3. Enable developer mode to verify Cache tags

### Steps to reproduce (*)

1. Enable varnish cache and make sure it is running
2. Open home page or any CMS page example: /about-us
3. Check Magento cache Tags.

### Expected result (*)
1. x-magento-tags: store,cms_b,cms_b_1,cms_b_footer_links_block,cms_p_5

### Actual result (*)
1. x-magento-tags: cat_c,cat_c_38,cat_c_20,cat_c_11,cat_c_3,cat_c_9,cat_c_37,store,cms_b,cms_b_1,cms_b_footer_links_block,cms_p_5

Due to this issue, if you update a product from top categories or run indexer to update product attribute or stock status it will wipe out full site page cache.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [s1] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by Jitheesh**
### Description (*)
https://github.com/magento/magento2/blob/bf4cdad2b1d0436be831c1f5c89c1adc77bfd1fc/app/code/Magento/PageCache/Model/Layout/LayoutPlugin.php#L88

Magento is using type comparison for determining varnish cache is enabled or not. But it returns false even if varnish is enabled.

This is because of following comparison; "2" === 2

Based on following code, I think it expected to skip adding tags for a top menu block
` $isEsiBlock = $block->getTtl() > 0;
                    if ($isVarnish && $isEsiBlock) {
                        continue;
                    }`

For Topmenu block we have TTL value so that $isEsiBlock become true and with my fix, $isVarnish will also become true.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#33465

### Manual testing scenarios (*)

1. Enable varnish cache and make sure it is running
2. Open home page or any CMS page example: /about-us
3. Check Magento cache Tags.
4. Cache tags won't include top category IDs after merging this PR

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by hostep** [comment]
Auwch, this was initially implemented correctly by https://github.com/magento/magento2/commit/b59dc722059f3a87716a41dc0f5ae62ec297fb69, but then reverted partially by https://github.com/magento/magento2/commit/1778c0f73570f19f9f368a18ff0808a6177f96e2

@gabrieldagama: is there news from the Fastly module around this, maybe they fixed something at their end to get their module working with this change?

**Comment by mrtuvn** [comment]
another pr supprised me with reverted code. Can't wait to hear any news from fastly

**Comment by Jitheesh** [comment]
Looks like these issues are also related 

- #30128
- #29806
- #25670

To replicate this issue try following steps

- Open a CMS page and make sure it is delivered from cache
- Edit price and stock status of a single product from top level category
- Run indexers
- Validate PURGE requests sent by Magento, it will contains tags like cat_c_<categoryid>
- Refresh CMS page, it will served from server instead of cache

**Comment by mrtuvn** [comment]
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/PageCache/Observer/ProcessLayoutRenderElement.php#L123
Seem we not check type here too

**Comment by ihor-sviziev** [comment]
Maybe it worth fixing the following code as well (convert to int)? https://github.com/magento/magento2/blob/bf4cdad2b1d0436be831c1f5c89c1adc77bfd1fc/app/code/Magento/PageCache/Model/Layout/LayoutPlugin.php#L88

**Comment by mrtuvn** [comment]
Found this one :)
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/PageCache/Observer/ProcessLayoutRenderElement.php#L123

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Could you fix the following code as well (convert to int)? https://github.com/magento/magento2/blob/bf4cdad2b1d0436be831c1f5c89c1adc77bfd1fc/app/code/Magento/PageCache/Model/Layout/LayoutPlugin.php#L88

**Comment by ihor-sviziev** [comment]
@hostep, unfortunately, we already had a case when conversion to int was reverted for some reason,  and I don't want to get it back. The missing return type can't guarantee that we will have int. I think it's better to add such conversion where it's used to make sure this issue won't appear again.

**Comment by Jitheesh** [comment]
> Could you fix the following code as well (convert to int)?
> 
> https://github.com/magento/magento2/blob/bf4cdad2b1d0436be831c1f5c89c1adc77bfd1fc/app/code/Magento/PageCache/Model/Layout/LayoutPlugin.php#L88

Instead of converting to int, can we skip  type comparison ? ie change === to ==

**Comment by mrtuvn** [comment]
Maybe we can check vanish enable like this 

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/PageCache/Observer/ProcessLayoutRenderElement.php#L123

**Comment by ihor-sviziev** [comment]
> > Could you fix the following code as well (convert to int)?
> > https://github.com/magento/magento2/blob/bf4cdad2b1d0436be831c1f5c89c1adc77bfd1fc/app/code/Magento/PageCache/Model/Layout/LayoutPlugin.php#L88
> 
> Instead of converting to int, can we skip type comparison ? ie change === to ==

TBH I don't think it's a good option, as not strict comparison might cause additional not expected issues.

**Comment by ihor-sviziev** [comment]
@Jitheesh, yes, please

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by vvuksan** [comment]
@hostep if something needs to be fixed in the Fastly module please open up an issue and we'll take a look.

**Comment by hostep** [comment]
@vvuksan: I'm not using the Fastly module, so I'm not interested in reporting bugs.
I'm just getting you involved because a bug was introduced in core Magento to prevent another bug to happen in the Fastly module, by this commit: https://github.com/magento/magento2/commit/1778c0f73570f19f9f368a18ff0808a6177f96e2 (it would have been better to fix this in the Fastly module than to revert that change in core Magento which introduced the problem that this PR is trying to fix).
If you are interested in this, please read through the entire thread, I think the info is clear enough to understand what is going on.

**Comment by mrtuvn** [comment]
@magento run Unit Tests, Functional Tests B2B

**Comment by mrtuvn** [comment]
Hi @Jitheesh Can you recheck Unit Tests

**Comment by mrtuvn** [comment]
@magento run Unit Tests

**Comment by ihor-sviziev** [comment]
FYI I reported an issue to Fastly extension https://github.com/fastly/fastly-magento2/issues/466

File: app/code/Magento/PageCache/Model/Config.php [hunk 1]
```diff
@@ -116,7 +116,7 @@ public function __construct(
     /**
      * Return currently selected cache type: built in or varnish
      *
-     * @return int
```
Comment (tanghongtanghong) [inline app/code/Magento/PageCache/Model/Config.php]: 看

**Comment by ihor-sviziev** [comment]
Update: Fastly has updated their module in https://github.com/fastly/fastly-magento2/pull/469.
I think we can introduce conversion to int inside the getType method. What do you think?

**Comment by mrtuvn** [comment]
That is quick change from fastly. Now from magento end i think update will be quick with less file changed

**Comment by mrtuvn** [comment]
Now  i think we have free to change from magento side

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @Jitheesh,
Fastly has fixed an issue on their side.
Could you add the conversion to int to the getType method, and remove in other places?

**Comment by ihor-sviziev** [comment]
Great! Thank you!

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@Jitheesh, Can you check unit test fail case related when change file app/code/Magento/PageCache/Observer/ProcessLayoutRenderElement.php

**Comment by engcom-Bravo** [comment]
@Jitheesh @ihor-sviziev and @mrtuvn  Thanks for your contribution!
@Jitheesh I am working on the requested changes.

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
Hi @mrtuvn

I have updated the PR with requested changes.
Can you please review PR.

**Comment by ihor-sviziev** [comment]
Looks like false positives. Let's try to re-run them

@magento run Functional Tests B2B, Functional Tests CE, Integration Tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9207](https://jira.corp.magento.com/browse/ENGCOM-9207) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@engcom-Alfa, any updates about testing of this PR?

**Comment by kestraly** [comment]
In the PR - Changes made to app/code/Magento/PageCache/Model/App/FrontController/BuiltinPlugin.php  https://github.com/magento/magento2/commit/f9b9ec1b2348970b49e18f3e7544424886b642b5 seems slow down loading times to uncached levels of speed at 1.5s with this change in place. 

I'm wondering if it's a combination of using Redis for FPC for this instance or not, so I'll do some experimenting.

 Reverting back to the original method of != in that file brings cached pages back down to 200ms.

**Pull request merged**